### PR TITLE
feat: Add helper function to verify bounds of timestamp search

### PR DESCRIPTION
**Pull request by OussamaSaoudi**
## What changes are proposed in this pull request?

 This PR introduces:
  - `timestamp_to_version`: Core function that converts a timestamp to a Delta version using binary search,
  supporting both GreatestLower (latest version at or before timestamp) and LeastUpper (first version at
  or after timestamp) bounds
  - get_timestamp_search_bounds: Determines the appropriate search strategy based on In-Commit Timestamp (ICT) enablement, handling the transition between file modification timestamps and ICT
  - TimestampSearchBounds: Enum representing the four possible search scenarios (exact match, ICT range,
  file modification range, or mixed)
  - Simplified LogHistoryError types with consolidated error variants

  The implementation correctly handles tables that transition from file modification timestamps to
  In-Commit Timestamps mid-history, ensuring searches are constrained to the appropriate timestamp domain.

**Stack:**
- PR #899 (this PR)
- PR #900

## How was this change tested?
Unit tests cover:
- Binary search with exact matches and no matches
- Duplicate value handling (first/last occurrence)
- Edge cases (empty arrays, out of range values)
- Error propagation from key functions

### CODE REVIEW

File: kernel/src/history_manager/mod.rs [hunk 1]
```diff
@@ -134,6 +147,152 @@ fn read_in_commit_timestamp(
...
+/// - `Bound::GreatestLower`: There is no commit whose timestamp is lower than the given `timestamp`.
+/// - `Bound::LeastUpper`: There is no commit whose timestamp is greater than the given `timestamp`.
+#[allow(unused)]
+fn timestamp_to_version(
+    snapshot: &Snapshot,
+    engine: &dyn Engine,
+    timestamp: Timestamp,
+    bound: Bound,
+) -> Result<Version, LogHistoryError> {
+    require!(timestamp >= 0, LogHistoryError::InvalidTimestamp(timestamp));
```
[the remark this answers is not in the document]
Comment (OussamaSaoudi) [reply kernel/src/history_manager/mod.rs]: Oh I didn't even make the connection that negative values can represent time before 1970! I'll fix that.

**Comment by OussamaSaoudi** [comment]
aside: I realized that the TimestampOutOfRangeError is now expensive to build without our cache of values. I can take that up as a followup

File: kernel/src/history_manager/error.rs [hunk 2]
```diff
@@ -41,3 +45,66 @@ pub enum LogHistoryError {
...
+    pub(crate) nearest_version: Version,
+    pub(crate) bound: Bound,
+    pub(crate) _read_ict: bool,
+}
+
+impl TimestampOutOfRangeError {
+    pub(crate) fn try_new<'a>(
+        commit_range: &'a [ParsedLogPath],
+        timestamp: Timestamp,
+        key_fun: impl Fn(&'a ParsedLogPath) -> Result<i64, LogHistoryError>,
```
[the remark this answers is not in the document]
Comment (OussamaSaoudi) [reply kernel/src/history_manager/error.rs]: good catch!

File: kernel/src/history_manager/error.rs [hunk 3]
```diff
@@ -42,3 +46,66 @@ pub enum LogHistoryError {
...
+    pub(crate) nearest_version: Version,
+    pub(crate) bound: Bound,
+    pub(crate) _read_ict: bool,
+}
+
+impl TimestampOutOfRangeError {
+    pub(crate) fn try_new<'a>(
+        commit_range: &'a [ParsedLogPath],
+        timestamp: Timestamp,
+        key_fun: impl Fn(&'a ParsedLogPath) -> Result<Timestamp, LogHistoryError>,
```
Comment (scovich) [inline kernel/src/history_manager/error.rs]: Remind me why these args need lifetimes? We don't keep any references that outlive the function call?
(I feel like I may have asked this before and received an answer, but I can't find it anywhere)

File: kernel/src/history_manager/error.rs [hunk 4]
```diff
@@ -42,3 +46,66 @@ pub enum LogHistoryError {
...
+            bound,
+            _read_ict: read_ict,
+        })
+    }
+}
+impl Display for TimestampOutOfRangeError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.bound {
+            Bound::LeastUpper => {
+                write!(
```
Comment (scovich) [inline kernel/src/history_manager/error.rs]: aside: I'm surprised fmt added curly braces here? I would have expected e.g.
```rust
Bound::LeastUpper => write!(
    ...
),
```

File: kernel/src/history_manager/mod.rs [hunk 5]
```diff
@@ -124,6 +136,146 @@ fn read_in_commit_timestamp(
...
+        return Ok(TimestampSearchBounds::FileModificationSearch);
+    };
+
+    // Helper function to get the of index at which ict was enabled if it is present in this
+    // LogSegment
+    let version_idx = || {
+        log_segment
+            .ascending_commit_files
+            .binary_search_by(|x| x.version.cmp(&ict_version))
+            .unwrap_or_else(|idx| idx)
```
Comment (scovich) [inline kernel/src/history_manager/mod.rs]: Maybe worth expanding the above comment to explain how we use the Err result here (why it's correct)?

Comment (scovich) [reply kernel/src/history_manager/mod.rs]: Also, why pull it out as a closure like this? Wouldn't it be just as compact if we fold it into its only callsite in the match below?

**Review by scovich — APPROVED** [review]
LGTM, few nits/questions to consider

File: kernel/src/history_manager/error.rs (hunk 3 above)
Comment (OussamaSaoudi) [reply kernel/src/history_manager/error.rs]: removed them 👍

File: kernel/src/history_manager/mod.rs (hunk 5 above)
Comment (OussamaSaoudi) [reply kernel/src/history_manager/mod.rs:82]: removed closure.
